python_lib/qss_compiler/exceptions.py

```diff
@@ -123,3 +126,60 @@ class QSSTargetUnsupportedOperation(QSSCompilerError):
 
 class OpenQASM3ParseFailure(QSSCompilerError):
     """Raised when a parser failure is received"""
+
+
+class OpenQASM3UnsupportedInput(QSSCompilerError):
+    """Raised when the input Openqasm 3 source uses semantics that are not supported by the
+    compiler."""
+
+
+# Map diagnostic categories to compiler exceptions
+category_exception_map = {
+    ErrorCategory.QSSCompilerSequenceTooLong: QSSCompilerSequenceTooLong,
+    ErrorCategory.QSSControlSystemResourcesExceeded: QSSControlSystemResourcesExceeded,
+    ErrorCategory.OpenQASM3ParseFailure: OpenQASM3ParseFailure,
+    ErrorCategory.QSSTargetUnsupportedOperation: QSSTargetUnsupportedOperation,
+    ErrorCategory.OpenQASM3UnsupportedInput: OpenQASM3UnsupportedInput,
+}
```

Comment:
The previous code was very redundant and polluted the `compile.py` with explicit if/elif matching and exception construction. I've consolidated this into a map. It would be even better if we could automatically generate these exceptions from diagnostics without having to manually add them.